File: backend/src/main/java/com/xml/processor/service/ProcessedFileService.java
package com.xml.processor.service;

import com.xml.processor.model.ProcessedFile;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import java.time.LocalDateTime;
import java.util.List;

public interface ProcessedFileService {
    @Cacheable(value = "processedFiles", key = "#id")
    ProcessedFile getProcessedFileById(Long id);

    @Cacheable(value = "processedFiles", key = "'all'")
    Page<ProcessedFile> getAllProcessedFiles(Pageable pageable);

    @Cacheable(value = "processedFiles", key = "'client_' + #clientId")
    Page<ProcessedFile> getProcessedFilesByClient(Long clientId, Pageable pageable);

    @Cacheable(value = "processedFiles", key = "'search_' + #fileName")
    Page<ProcessedFile> searchProcessedFiles(String fileName, Pageable pageable);

    @Cacheable(value = "processedFiles", key = "'status_' + #status")
    Page<ProcessedFile> getProcessedFilesByStatus(String status, Pageable pageable);

    @Cacheable(value = "processedFiles", key = "'date_range_' + #startDate + '_' + #endDate")
    Page<ProcessedFile> getProcessedFilesByDateRange(
        LocalDateTime startDate,
        LocalDateTime endDate,
        Pageable pageable
    );

    @Cacheable(value = "processedFiles", key = "'client_status_' + #clientId + '_' + #status")
    Page<ProcessedFile> getProcessedFilesByClientAndStatus(
        Long clientId,
        String status,
        Pageable pageable
    );

    @Cacheable(value = "processedFiles", key = "'client_date_' + #clientId + '_' + #startDate + '_' + #endDate")
    Page<ProcessedFile> getProcessedFilesByClientAndDateRange(
        Long clientId,
        LocalDateTime startDate,
        LocalDateTime endDate,
        Pageable pageable
    );

    @CacheEvict(value = "processedFiles", allEntries = true)
    ProcessedFile createProcessedFile(ProcessedFile processedFile);

    @CacheEvict(value = "processedFiles", key = "#id")
    ProcessedFile updateProcessedFile(Long id, ProcessedFile processedFile);

    @CacheEvict(value = "processedFiles", key = "#id")
    void deleteProcessedFile(Long id);

    Page<ProcessedFile> getProcessedFiles(int page, int size, String sortBy, String sortDirection, String searchTerm, String status, LocalDateTime startDate, LocalDateTime endDate);
    Page<ProcessedFile> getProcessedFilesByClient(Long clientId, int page, int size, String sortBy, String sortDirection);
    Page<ProcessedFile> getProcessedFilesByClientAndStatus(Long clientId, String status, int page, int size, String sortBy, String sortDirection);
    Page<ProcessedFile> getProcessedFilesByClientAndDateRange(Long clientId, LocalDateTime startDate, LocalDateTime endDate, int page, int size, String sortBy, String sortDirection);
    List<ProcessedFile> getProcessedFilesByStatus(String status);
} 

File: frontend/src/services/processedFileService.ts
import axios from 'axios';
import { ProcessedFile, PageResponse } from '../types';

const API_URL = process.env.REACT_APP_API_URL || 'http://localhost:8080/api';

export const processedFileService = {
  getAllProcessedFiles: async (
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc',
    fileNameFilter?: string,
    statusFilter?: string,
    startDate?: string,
    endDate?: string
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    if (fileNameFilter) {
      params.append('fileNameFilter', fileNameFilter);
    }
    
    if (statusFilter) {
      params.append('statusFilter', statusFilter);
    }
    
    if (startDate) {
      params.append('startDate', startDate);
    }
    
    if (endDate) {
      params.append('endDate', endDate);
    }
    
    const response = await axios.get<PageResponse<ProcessedFile>>(`${API_URL}/processed-files?${params.toString()}`);
    return response.data;
  },

  getProcessedFileById: async (id: number): Promise<ProcessedFile> => {
    const response = await axios.get<ProcessedFile>(`${API_URL}/processed-files/${id}`);
    return response.data;
  },

  createProcessedFile: async (processedFileData: Omit<ProcessedFile, 'id'>): Promise<ProcessedFile> => {
    const response = await axios.post<ProcessedFile>(`${API_URL}/processed-files`, processedFileData);
    return response.data;
  },

  updateProcessedFile: async (id: number, processedFileData: Partial<ProcessedFile>): Promise<ProcessedFile> => {
    const response = await axios.put<ProcessedFile>(`${API_URL}/processed-files/${id}`, processedFileData);
    return response.data;
  },

  deleteProcessedFile: async (id: number): Promise<void> => {
    await axios.delete(`${API_URL}/processed-files/${id}`);
  },

  getProcessedFilesByClient: async (
    clientId: number,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await axios.get<PageResponse<ProcessedFile>>(`${API_URL}/processed-files/client/${clientId}?${params.toString()}`);
    return response.data;
  },

  searchProcessedFiles: async (
    fileName: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('fileName', fileName);
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await axios.get<PageResponse<ProcessedFile>>(`${API_URL}/processed-files/search?${params.toString()}`);
    return response.data;
  },

  getProcessedFilesByStatus: async (
    status: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await axios.get<PageResponse<ProcessedFile>>(`${API_URL}/processed-files/status/${status}?${params.toString()}`);
    return response.data;
  },

  getProcessedFilesByDateRange: async (
    startDate: string,
    endDate: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('startDate', startDate);
    params.append('endDate', endDate);
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await axios.get<PageResponse<ProcessedFile>>(`${API_URL}/processed-files/date-range?${params.toString()}`);
    return response.data;
  },

  getProcessedFilesByClientAndStatus: async (
    clientId: number,
    status: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await axios.get<PageResponse<ProcessedFile>>(`${API_URL}/processed-files/client/${clientId}/status/${status}?${params.toString()}`);
    return response.data;
  },

  getProcessedFilesByClientAndDateRange: async (
    clientId: number,
    startDate: string,
    endDate: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('startDate', startDate);
    params.append('endDate', endDate);
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await axios.get<PageResponse<ProcessedFile>>(`${API_URL}/processed-files/client/${clientId}/date-range?${params.toString()}`);
    return response.data;
  }
}; 